bap-vibes/src/linear_ssa.ml

```diff
@@ -188,53 +193,72 @@ let linearize_blk (blk : blk term) : blk term linear =
      congruence between linear SSA vars. *)
   Blk.create ~phis:[] ~defs ~jmps ~tid:(Term.tid blk) ()
 
-(** [compute_liveness] produces a linear_ssa-ified live variable map. Each variable
-    get the prefix from it's block *)
-let compute_liveness (sub : sub term) : Data.ins_outs Tid.Map.t =
-  let liveness = Sub.compute_liveness sub in
-  let blks = Term.enum blk_t sub in
-  let ins_outs_map = Seq.map blks ~f:(fun blk ->
-    let tid = Term.tid blk in
-    let outs = Graphlib.Std.Solution.get liveness tid in
-    let prefix = prefix_from blk in
-    (* Delete phis because they "define" variables in way we don't want *)
-    let blk_no_phi = Blk.create
-      ~defs:(Seq.to_list @@ Term.enum def_t blk)
-      ~jmps:(Seq.to_list @@ Term.enum jmp_t blk) () in
-    (* The ins are the outs minus the variables defined in the block unioned with
-       any variable occuring free in the blocks (possibly last occurences so possdibly not
-       live at end of block). *)
-    let ins = Var.Set.filter outs ~f:(fun var -> not (Blk.defines_var blk_no_phi var)) in
-    let ins = Var.Set.union (Blk.free_vars blk_no_phi) ins in
-    let outs = Var.Set.map outs ~f:(fun var -> linearize ~prefix var) in
-    let ins = Var.Set.map ins ~f:(fun var -> linearize ~prefix var) in
-    let ins_outs : Data.ins_outs = {ins; outs} in
-    tid, ins_outs)
-  in
-  Tid.Map.of_sequence_exn ins_outs_map
+(* Produce a linear_ssa-ified live variable map. Each variable
+   get the prefix from it's block *)
+let compute_liveness
+    (sub : sub term) : (blk term list * Data.ins_outs Tid.Map.t) KB.t =
+  let open KB.Let in
+  let liveness = Live.compute sub in
+  let blks = Term.enum blk_t sub |> Seq.to_list in
+  (* Map each block to a list of pseudo definitions according to the
+     phi nodes we discovered. *)
+  let phi_map =
```

Comment:
It would help to have a type annotation here. This is `(var * exp) Tid.Map.t` ?

Reply:
And the mapping is block tids of precessors yes?

Reply:
Sure, and yes to the second comment.